publish-docs-to-s3 fails on tags cut before #66100

## Issue

`.github/workflows/publish-docs-to-s3.yml` runs
`python3 dev/registry/derive_wave_providers.py` unconditionally in the
**Build Info** job. The script was introduced by #66100 (commit
`b5e9bae6dc`, 2026-04-30). Any release tag cut before that commit does
not contain the script, so the docs publish workflow fails at this
step when invoked on those tags.

Hit while publishing docs for `airflow-ctl/0.1.4` (cut from
`0.1.4rc3` on 2026-04-22). The Build Info job failed with:

```
python3: can't open file '.../dev/registry/derive_wave_providers.py': [Errno 2] No such file or directory
##[error]Process completed with exit code 2.
```

Failed runs:
- https://github.com/apache/airflow/actions/runs/25392730378
  (workflow on `main`, `--ref airflow-ctl/0.1.4`)
- https://github.com/apache/airflow/actions/runs/25392898972
  (same, with `--apply-commits b5e9bae6dc` — `--apply-commits` runs in
  a later job, doesn't reach the Build Info failure point)

## Workaround

Pass `--workflow-branch <release-tag>` so the workflow YAML at the
tag is used (the workflow at the tag predates the registry-derive
step):

```
breeze workflow-run publish-docs \
  --ref airflow-ctl/0.1.4 \
  --workflow-branch airflow-ctl/0.1.4 \
  apache-airflow-ctl
```

That run: https://github.com/apache/airflow/actions/runs/25393124890

## Fix options

1. Wrap the registry-derive step in
   `if: hashFiles('dev/registry/derive_wave_providers.py') != ''`
   (or guard the `run:` block with `[ -f ... ] && python3 ...`).
   Lowest-churn fix; lets old tags publish docs with the current
   workflow definition.
2. Skip the registry-derive step entirely for non-provider
   distributions (`airflow-ctl`, `helm-chart`, etc.) — the registry
   rebuild is provider-targeted, so it's a no-op for these anyway.
3. Cherry-pick the script onto release branches during release prep.
   Higher friction; doesn't help for tags already cut.

## Scope

Any tag predating `b5e9bae6dc` on the workflow branch — anyone
republishing docs for an older release (e.g. a fix backport on an
older provider release) will hit this. Workaround is universal but
adds a release-manager footgun.

---
Drafted-by: Claude Code (Opus 4.7); reviewed by @potiuk before posting

--- a/.github/workflows/publish-docs-to-s3.yml
+++ b/.github/workflows/publish-docs-to-s3.yml
@@ -127,7 +127,39 @@ jobs:
         env:
           INCLUDE_DOCS: ${{ inputs.include-docs }}
           REF: ${{ inputs.ref }}
-        run: python3 dev/registry/derive_wave_providers.py
+        shell: bash
+        # The registry rebuild is provider-targeted. For non-provider
+        # distributions (apache-airflow, airflow-ctl, helm-chart, etc.) the
+        # update-registry job below is a no-op anyway, so skip the derive
+        # step — that lets release tags cut before
+        # `dev/registry/derive_wave_providers.py` was added (#66100) still
+        # publish docs without the workflow tripping on the missing file.
+        # Tracked at #66436.
+        run: |
+          # Keep the non-provider list in sync with NON_PROVIDER_TOKENS in
+          # dev/registry/derive_wave_providers.py.
+          NON_PROVIDER_TOKENS=("apache-airflow" "apache-airflow-ctl" "apache-airflow-task-sdk" "helm-chart" "docker-stack")
+          has_provider=false
+          for token in $INCLUDE_DOCS; do
+            is_non_provider=false
+            for np in "${NON_PROVIDER_TOKENS[@]}"; do
+              if [[ "$token" == "$np" ]]; then
+                is_non_provider=true
+                break
+              fi
+            done
+            if ! $is_non_provider; then
+              has_provider=true
+              break
+            fi
+          done
+          if ! $has_provider; then
+            echo "Non-provider distribution(s) only ('${INCLUDE_DOCS}') — skipping registry-derive."
+            echo "registry-providers=" >> "${GITHUB_OUTPUT}"
+            echo "registry-full-build=false" >> "${GITHUB_OUTPUT}"
+            exit 0
+          fi
+          python3 dev/registry/derive_wave_providers.py
       - name: "Input parameters summary"
         shell: bash
         id: parameters

